homeassistant/components/mobile_app/helpers.py

```diff
@@ -0,0 +1,103 @@
+"""Helpers for mobile_app."""
+import logging
+import json
+from typing import Callable, Dict, Tuple
+
+from aiohttp.web import Response
+
+from homeassistant.core import Context
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import (ATTR_APP_DATA, ATTR_APP_ID, ATTR_APP_NAME,
+                    ATTR_APP_VERSION, ATTR_DELETED_IDS, ATTR_DEVICE_NAME,
+                    ATTR_MANUFACTURER, ATTR_MODEL, ATTR_OS_VERSION,
+                    ATTR_REGISTRATIONS, ATTR_SUPPORTS_ENCRYPTION,
+                    CONF_USER_ID, DOMAIN)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_cipher() -> Tuple[int, Callable]:
+    """Return decryption function and length of key.
+
+    Async friendly.
+    """
+    from nacl.secret import SecretBox
+    from nacl.encoding import Base64Encoder
+
+    def decrypt(ciphertext, key):
+        """Decrypt ciphertext using key."""
+        return SecretBox(key).decrypt(ciphertext, encoder=Base64Encoder)
+    return (SecretBox.KEY_SIZE, decrypt)
+
+
+def _decrypt_payload(key: str, ciphertext: str) -> Dict[str, str]:
+    """Decrypt encrypted payload."""
+    try:
+        keylen, decrypt = get_cipher()
+    except OSError:
+        _LOGGER.warning(
+            "Ignoring encrypted payload because libsodium not installed")
+        return None
+
+    if key is None:
+        _LOGGER.warning(
+            "Ignoring encrypted payload because no decryption key known")
+        return None
+
+    key = key.encode("utf-8")
+    key = key[:keylen]
+    key = key.ljust(keylen, b'\0')
+
+    try:
+        message = decrypt(ciphertext, key)
+        message = json.loads(message.decode("utf-8"))
+        _LOGGER.debug("Successfully decrypted mobile_app payload")
+        return message
+    except ValueError:
+        _LOGGER.warning("Ignoring encrypted payload because unable to decrypt")
+        return None
+
+
+def device_context(device: Dict) -> Context:
+    """Generate a context from a request."""
+    return Context(user_id=device[CONF_USER_ID])
+
+
+def empty_okay_response(headers: Dict = None) -> Response:
+    """Return a Response with empty JSON object and a 200."""
+    return Response(body='{}', status=200, content_type='application/json',
```

Comment:
Why set a body and content to JSON ?

Reply:
Doing Response(status=200) returns a 200 with body “None” and Content-Type application/octet-stream. Would rather be returning empty JSON objects so response can be expected to always be JSON.

Reply:
well the downside of that is that we give some extra info about the response that other webhooks don't.

Reply:
Agreed to keep it as is.